spark_log_parser/parsing_models/application_model.py

```diff
@@ -121,79 +126,51 @@ def __init__(self, eventlogpath, bucket=None, stdoutpath=None, debug=False):  #
 
                     # print("Stage ids: %s" % stage_ids)
                     for stage_id in stage_ids:
-                        if stage_id not in self.jobs_for_stage:
-                            self.jobs_for_stage[stage_id] = [job_id]
-                        else:
-                            self.jobs_for_stage[stage_id].append(job_id)
+                        self.jobs_for_stage[stage_id].append(job_id)
 
                 elif event_type == "SparkListenerJobEnd":
                     job_id = json_data["Job ID"]
                     self.jobs[job_id].completion_time = json_data["Completion Time"] / 1000
                     self.jobs[job_id].result = json_data["Job Result"]["Result"]
+
                 elif event_type == "SparkListenerTaskEnd":
-                    stage_id = json_data["Stage ID"]
-                    # Add the event to all of the jobs that depend on the stage.
-                    for jid in self.jobs_for_stage[stage_id]:
-                        self.jobs[jid].add_event(json_data, True)
+                    if "Task Metrics" in json_data:
+                        task = TaskModel(json_data, True)
+                        self.tasks.append(task)
 
                 elif event_type == "SparkListenerStageSubmitted":
+                    stage_info = json_data["Stage Info"]
 
-                    # stages may not be executed exclusively from one job
-                    stage_id = json_data["Stage Info"]["Stage ID"]
-
-                    if (stage_id not in self.jobs_for_stage) and (not debug):
-                        raise UrgentEventValidationException(
-                            missing_event=f"Job Start for Stage {stage_id}"
-                        )
-
-                    if "Submission Time" not in json_data["Stage Info"]:
+                    if "Submission Time" not in stage_info:
                         # PROD-426 Submission Time key may be missing from stages that
                         # don't get submitted. There is usually a StageCompleted event
-                        # shortly after.
+                        # shortly after. This may happen when stages fail
                         continue
 
-                    for job_id in self.jobs_for_stage[stage_id]:
```

Comment:
Having 1 `StageModel` removes the need to do this iteration each time we are updating some data for a Stage

Reply:
We also discussed the fact that failed stages should likely be distinct from successful ones, and that we can discriminate using the `Attempt ID` that we receive from a Stage. I think that moving towards this 1 `StageModel` per ID (whether that is just `stage_id`, or `(stage_id, attempt_id)`) should make that refactoring a bit easier as well.

Reply:
Nice! Besides enabling the out-of-order reading, separating the tasks/stages/jobs from the get-go is just so much cleaner and easier to follow.